test/cmd/images.sh

```diff
@@ -163,14 +175,40 @@ os::cmd::expect_success 'oc tag mysql:latest tagtest:zzz tagtest2:zzz --alias'
 os::cmd::expect_success_and_text "oc get is/tagtest --template='{{(index .spec.tags 8).from.kind}}'" 'ImageStreamTag'
 os::cmd::expect_success_and_text "oc get is/tagtest2 --template='{{(index .spec.tags 0).from.kind}}'" 'ImageStreamTag'
 
+# test copying tags
 os::cmd::expect_success 'oc tag registry-1.docker.io/openshift/origin:v1.0.4 newrepo:latest'
 os::cmd::expect_success_and_text "oc get is/newrepo --template='{{(index .spec.tags 0).from.kind}}'" 'DockerImage'
 os::cmd::try_until_success 'oc get istag/mysql:5.5'
+# default behavior is to copy the current image, but since this is an external image we preserve the dockerImageReference
 os::cmd::expect_success 'oc tag mysql:5.5 newrepo:latest'
 os::cmd::expect_success_and_text "oc get is/newrepo --template='{{(index .spec.tags 0).from.kind}}'" 'ImageStreamImage'
+os::cmd::expect_success_and_text "oc get is/newrepo --template='{{(index .status.tags 0 \"items\" 0).dockerImageReference}}'" '^mysql@sha256:'
+# aliases set the spec tag to be a reference to the originating stream
 os::cmd::expect_success 'oc tag mysql:5.5 newrepo:latest --alias'
 os::cmd::expect_success_and_text "oc get is/newrepo --template='{{(index .spec.tags 0).from.kind}}'" 'ImageStreamTag'
-
+os::cmd::expect_success_and_text "oc get is/newrepo --template='{{(index .status.tags 0 \"items\" 0).dockerImageReference}}'" '^mysql@sha256:'
+# when copying a tag that points to the internal registry, update the docker image reference
+os::cmd::expect_success "oc tag test:new newrepo:direct"
+os::cmd::expect_success_and_text 'oc get istag/newrepo:direct -o jsonpath={.image.dockerImageReference}' "/$project/newrepo@sha256:"
+# test references
+os::cmd::expect_success 'oc tag mysql:5.5 reference:latest --alias --reference'
+os::cmd::expect_success 'oc tag mysql:5.5 reference:other --reference'
+os::cmd::expect_success_and_text "oc get is/reference --template='{{(index .spec.tags 0).from.kind}}'" 'ImageStreamTag'
+os::cmd::expect_success_and_text "oc get is/reference --template='{{(index .spec.tags 0).reference}}'" 'true'
+os::cmd::expect_success_and_text "oc get is/reference --template='{{(index .spec.tags 1).from.kind}}'" 'ImageStreamImage'
```

Comment:
Comments addressed PTAL

On Thu, Jun 9, 2016 at 11:22 AM, Clayton Coleman <ccoleman@redhat.com>
wrote:

> It's only a map internally.  Externally it's an array.  And it's sorted.
>
> On Thu, Jun 9, 2016 at 5:35 AM, Michal Minar <notifications@github.com>
> wrote:
>
>> In test/cmd/images.sh
>> <https://github.com/openshift/origin/pull/9238#discussion_r66409988>:
>>
>> >  os::cmd::expect_success 'oc tag mysql:5.5 newrepo:latest --alias'
>> >  os::cmd::expect_success_and_text "oc get is/newrepo --template='{{(index .spec.tags 0).from.kind}}'" 'ImageStreamTag'
>> > -
>> > +os::cmd::expect_success_and_text "oc get is/newrepo --template='{{(index .status.tags 0 \"items\" 0).dockerImageReference}}'" '^mysql@sha256:'
>> > +# when copying a tag that points to the internal registry, update the docker image reference
>> > +os::cmd::expect_success "oc tag test:new newrepo:direct"
>> > +os::cmd::expect_success_and_text 'oc get istag/newrepo:direct -o jsonpath={.image.dockerImageReference}' "/$project/newrepo@sha256:"
>> > +# test references
>> > +os::cmd::expect_success 'oc tag mysql:5.5 reference:latest --alias --reference'
>> > +os::cmd::expect_success 'oc tag mysql:5.5 reference:other --reference'
>> > +os::cmd::expect_success_and_text "oc get is/reference --template='{{(index .spec.tags 0).from.kind}}'" 'ImageStreamTag'
>> > +os::cmd::expect_success_and_text "oc get is/reference --template='{{(index .spec.tags 0).reference}}'" 'true'
>> > +os::cmd::expect_success_and_text "oc get is/reference --template='{{(index .spec.tags 1).from.kind}}'" 'ImageStreamImage'
>>
>> I don't really understand how a map[string]... can be indexed by
>> integer. And even if the evaluator iterated over keys with index, you may
>> get different key each time since they're unordered.
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/openshift/origin/pull/9238/files/b7b4a40b35a57e46528ad9efb1f26e309b8bc8fd..7d726a83f37add9d855d96696a30d8e2d296a314#r66409988>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe/ABG_p-hb3nCXlLedE8UsFKtsGSOmrN4kks5qJ95ngaJpZM4IxjR3>
>> .
>>
>
>